libraries/botframework-schema/src/activityEx.ts

```diff
@@ -0,0 +1,405 @@
+            return true;
+        }
+
+        if(source.channelData !== undefined) {
+            return true;
+        }
+
+        return false;
+    }
+
+  /**
+   * Resolves the mentions from the entities of the source activity.
+   * @param source The source activity.
+   * @returns The array of mentions; or an empty array, if none are found.
+   * @remarks This method is only intended for use with a message activity, where the Activity Type is set to Message.
+   * @see cref="entities" .
+   * @see cref="mention" .
+   */
+    export function getMentions(source: Partial<Activity>) : Mention[] {
+        return source.entities.filter(x => x.type.toLowerCase() === 'mention')
+        .map(e => e as Mention);
+    }
+
+  /**
+   * Creates a ConversationReference based on the source activity.
+   * @param source The source activity.
+   * @returns A conversation reference for the conversation that contains the activity.
+   */
+    export function getConversationReference(source: Partial<Activity>) : Partial<ConversationReference> {
+        return {
+            activityId: source.id,
+            user: source.from,
+            bot: source.recipient,
+            conversation: source.conversation,
+            channelId: source.channelId,
+            locale: source.locale,
+            serviceUrl: source.serviceUrl
+        } as ConversationReference;
+    }
+
+  /**
+   * Determines if the Activity was sent via an Http/Https connection or Streaming.
+   * This can be determined by looking at the ServiceUrl property:
+   * (1) All channels that send messages via http/https are not streaming
+   * (2) Channels that send messages via streaming have a ServiceUrl that does not begin with http/https.
+   * @param source The source activity.
+   * @returns True if the Activity was originate from a streaming connection.
+   */
+    export function isFromStreamingConnection(source: Partial<Activity>) : boolean {
+        if (source.serviceUrl !== undefined) {
+            const isHttp: boolean = source.serviceUrl.toLowerCase().startsWith("http");
+            return !isHttp;
+        }
+        return false;
+    }
+
+  /**
+   * Indicates whether this activity is of a specified activity type.
+   * @param source The source activity.
+   * @param activityType The activity type to check for.
+   * @returns True if the activity is of the specified activity type; otherwise, false.
+   */
+    export function isActivity(source:Partial<Activity>, activityType: string): boolean {
+       /*
+        * NOTE: While it is possible to come up with a fancy looking "one-liner" to solve
+        * this problem, this code is purposefully more verbose due to optimizations.
+        *
+        * This main goal of the optimizations was to make zero allocations because it is called
+        * by all of the .asXXXActivity methods which are used in a pattern heavily upstream to
+        * "pseudo-cast" the activity based on its type.
+        */

```

Comment:
Is this true? Or is this a blind copy-paste from the .NET implementation?

Reply:
Hi @stevengum, this was copied from .NET implementation but we consider it applies here as well, as the methods that call this one were almost identically migrated from .NET. We updated the comment to clarify this